### PR TITLE
Implement support for Quantity search parameters in the Hibernate Search index

### PR DESCRIPTION
**Pull request by michaelabuckley**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java [hunk 1]
```diff
@@ -453,4 +463,121 @@ private PredicateFinalStep generateDateInstantSearchTerms(String theSearchParamN
...
+						.field(theFieldName)
+						.between(theValue-approxTolerance, theValue+approxTolerance));
+				break;
+
+			// searches for resource quantity between passed param value +/- 5%
+			case EQUAL:
+				theQuantityTerms.must(
+					myPredicateFactory.range()
+						.field(theFieldName)
+						.between(theValue-defaultTolerance, theValue+defaultTolerance));
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java]: Good catch! That was fixed in later commits

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java [hunk 2]
```diff
@@ -453,4 +465,126 @@ private PredicateFinalStep generateDateInstantSearchTerms(String theSearchParamN
...
+
+				String fieldPath = NESTED_SEARCH_PARAM_ROOT + "." + theSearchParamName + "." + QTY_PARAM_NAME;
+				BooleanPredicateClausesStep<?> quantityTerms = myPredicateFactory.bool();
+
+				QuantityParam canonicalQty = UcumServiceUtil.toCanonicalQuantityOrNull(qtyParam);
+				if (canonicalQty != null) {
+					String valueFieldPath = fieldPath + "." + QTY_VALUE_NORM;
+					setPrefixedQuantityPredicate(quantityTerms, activePrefix, canonicalQty, valueFieldPath);
+					quantityTerms.must(myPredicateFactory.match()
+						.field(fieldPath + "." + QTY_CODE_NORM)
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java:494]: They are different. One for VALUE and the other for CODE

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java [hunk 3]
```diff
@@ -453,4 +465,126 @@ private PredicateFinalStep generateDateInstantSearchTerms(String theSearchParamN
...
+				break;
+
+			// searches for resource quantity not < param value
+			case GREATERTHAN_OR_EQUALS:
+				theQuantityTerms.mustNot(
+					myPredicateFactory.range()
+						.field(valueFieldPath)
+						.lessThan(value));
+				break;
+
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java:561]: I see your point, however that is not our inconsistency, as that code follows the specification.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java [hunk 4]
```diff
@@ -453,4 +465,126 @@ private PredicateFinalStep generateDateInstantSearchTerms(String theSearchParamN
...
+	 * Differences with DB search:
+	 *  _ is not all-normalized-or-all-not. Each parameter is applied on quantity or normalized quantity depending on UCUM fitness
+	 *  _ respects ranges for equal and approximate qualifiers
+	 *
+	 * Strategy: For each parameter, if it can be canonicalized, it is, and used against 'normalized-value-quantity' index
+	 * 	otherwise it is applied as-is to 'value-quantity'
+	 */
+	public void addQuantityUnmodifiedSearch(String theSearchParamName, List<List<IQueryParameterType>> theQuantityAndOrTerms) {
+
+		for (List<IQueryParameterType> nextAnd : theQuantityAndOrTerms) {
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java:480]: The parameter is a list of lists where the external list represents "and" terms and the internal list "or" terms

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedLuceneIndexData.java [hunk 5]
```diff
@@ -84,6 +90,8 @@ public void writeIndexElements(DocumentElement theDocument) {
 		mySearchParamStrings.forEach(ifNotContained(indexWriter::writeStringIndex));
 		mySearchParamTokens.forEach(ifNotContained(indexWriter::writeTokenIndex));
 		mySearchParamLinks.forEach(ifNotContained(indexWriter::writeReferenceIndex));
+//		mySearchParamQuantities.forEach(ifNotContained(indexWriter::writeQuantityIndex));
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedLuceneIndexData.java:93]: The reason is that we want to receive the whole entry collection for each invocation.

I commented that in the code and removed the commented out old code

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/HibernateSearchIndexWriter.java [hunk 6]
```diff
@@ -97,4 +117,38 @@ public void writeDateIndex(String theSearchParam, DateSearchIndexData theValue)
...
+			nestedQtyNode.addValue(QTY_VALUE, theValue.getValue());
+
+			if ( ! myModelConfig.getNormalizedQuantitySearchLevel().storageOrSearchSupported()) { return; }
+
+			//-- convert the value/unit to the canonical form if any
+			Pair canonicalForm = UcumServiceUtil.getCanonicalForm(theValue.getSystem(),
+				BigDecimal.valueOf(theValue.getValue()), theValue.getCode());
+			if (canonicalForm == null) { return; }
+
+			BigDecimal canonicalValue = BigDecimal.valueOf(Double.parseDouble(canonicalForm.getValue().asDecimal()));
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/HibernateSearchIndexWriter.java:141]: Good point! Actually the use of BD there was a remain of a refactoring, so I just changed it for a double.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (JasonRoberts-smile) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java:480]: Well, that's my point. I don't see where you are handling the or-ness of the inner list. If the query says `x = (5 or 8)`, as far as I can tell, the code will produce one predicate x.matches(5) and one predicate x.matches(8) and the or relationship is lost.

But when we start getting into inequality operators, supporting or doesn't even make sense. `x > (5 or 8)` is at best redundant, and at worst just wrong. I'll suggest one more time that you consider the way the date search handles it, but I won't insist.

**Review by michaelabuckley — COMMENTED** [review]
It looks like you caught the or problem.

A couple of minor suggestions about how to implement lt/gt.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java [hunk 7]
```diff
@@ -453,4 +465,126 @@ private PredicateFinalStep generateDateInstantSearchTerms(String theSearchParamN
...
+			for (IQueryParameterType paramType : nextAnd) {
+
+				QuantityParam qtyParam = QuantityParam.toQuantityParam(paramType);
+				ParamPrefixEnum activePrefix = qtyParam.getPrefix() == null ? ParamPrefixEnum.EQUAL : qtyParam.getPrefix();
+
+				String fieldPath = NESTED_SEARCH_PARAM_ROOT + "." + theSearchParamName + "." + QTY_PARAM_NAME;
+				BooleanPredicateClausesStep<?> quantityTerms = myPredicateFactory.bool();
+
+				QuantityParam canonicalQty = UcumServiceUtil.toCanonicalQuantityOrNull(qtyParam);
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneClauseBuilder.java]: Sure. Changing that.

**Review by jaison-b — APPROVED** [review]

**Pull request merged**